src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LoadedProject.cs

```diff
@@ -95,6 +101,22 @@ private void ProjectFileChangeContext_FileChanged(object? sender, string filePat
         NeedsReload?.Invoke(this, EventArgs.Empty);
     }
 
+    private void AssetsFileChangeContext_FileChanged(object? sender, string filePath)
+    {
+        // We only want to trigger design time build if the assets file content actually changed from the last time this handler was called.
+        // Sometimes we can get a change event where no content changed (e.g. for a failed restore).
+        // In such cases, proceeding with design-time build can put us in a restore loop (since the design-time build notices that assets are missing).
+        using var assetsFileStream = File.OpenRead(filePath);
+        var sourceText = SourceText.From(assetsFileStream);
+        var contentHash = sourceText.GetContentHash();
```

Comment:
You could just use https://github.com/dotnet/roslyn/blob/6a3a568f4849590011f6717f813e93eb2b43e976/src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs?plain=1#L71C5-L71C49 and directly pass the stream so we avoid having to realize the full thing. That also returns our existing Checksum object directly which will simplify your code a bit.